adapters/applogy/applogy.go

```diff
@@ -0,0 +1,157 @@
+package applogy
+
+import (
+	"encoding/json"
+	"errors"
+	"net/http"
+	"strconv"
+
+	"github.com/mxmCherry/openrtb"
+	"github.com/prebid/prebid-server/adapters"
+	"github.com/prebid/prebid-server/errortypes"
+	"github.com/prebid/prebid-server/openrtb_ext"
+)
+
+type ApplogyAdapter struct {
+	endpoint string
+}
+
+func (a *ApplogyAdapter) MakeRequests(request *openrtb.BidRequest, _ *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+	headers.Add("Accept", "application/json")
+	impressions := request.Imp
+	totalImpressions := len(impressions)
+	result := make([]*adapters.RequestData, 0, totalImpressions)
+	errs := make([]error, 0, totalImpressions)
+
+	for i, impression := range impressions {
+		if impression.Banner == nil && impression.Video == nil && impression.Native == nil {
+			errs = append(errs, &errortypes.BadInput{
+				Message: "Applogy only supports banner, video or native ads",
+			})
+			continue
+		}
+		if impression.Banner != nil {
+			banner := impression.Banner
+			if banner.W == nil || banner.H == nil || *banner.W == 0 || *banner.H == 0 {
+				if len(banner.Format) == 0 {
+					errs = append(errs, &errortypes.BadInput{
+						Message: "banner size information missing",
+					})
+					continue
+				}
+				format := banner.Format[0]
+				banner.W = &format.W
+				banner.H = &format.H
+			}
+		}
```

Comment:
What happens if `impression.Banner == nil` but `impression.Video != nil` or `impression.Native != nil`? Will we perform no checks on any of them?

Reply:
Only banner has format, video and native no need to check.